File: src/tts/ttsVoiceSelection.ts
import type { TtsVoice } from './ttsTypes';

const normalizeLanguageTag = (value: string): string => value.replace(/_/g, '-').toLowerCase();

const getLanguageCandidates = (value: string): string[] => {
  const normalized = normalizeLanguageTag(value);
  const [primary, region] = normalized.split('-');

  if (primary === 'he' || primary === 'iw') {
    return region ? [`he-${region}`, `iw-${region}`, 'he', 'iw'] : ['he', 'iw'];
  }

  return region ? [normalized, primary] : [primary];
};

const PLACEHOLDER_VOICE_PATTERN = /[-_]language$/;

const getVoicePriority = (voice: TtsVoice): number => {
  const identifier = voice.identifier.toLowerCase();
  let score = 0;

  if (identifier.includes('local')) {
    score += 4;
  }

  if (voice.quality === 'Enhanced') {
    score += 2;
  }

  if (identifier.includes('default')) {
    score += 1;
  }

  if (identifier.includes('network')) {
    score -= 1;
  }

  if (PLACEHOLDER_VOICE_PATTERN.test(identifier)) {
    score -= 5;
  }

  return score;
};

const pickPreferredVoice = (voices: TtsVoice[]): TtsVoice | undefined => {
  let preferredVoice: TtsVoice | undefined;
  let preferredScore = Number.NEGATIVE_INFINITY;

  for (const voice of voices) {
    const score = getVoicePriority(voice);

    if (!preferredVoice || score > preferredScore) {
      preferredVoice = voice;
      preferredScore = score;
    }
  }

  return preferredVoice;
};

export const findBestVoiceForLanguage = (
  voices: TtsVoice[],
  language: string,
): TtsVoice | undefined => {
  for (const candidate of getLanguageCandidates(language)) {
    const matchedVoice = pickPreferredVoice(
      voices.filter((voice) => normalizeLanguageTag(voice.language) === candidate),
    );

    if (matchedVoice) {
      return matchedVoice;
    }
  }

  return undefined;
};

File: android/app/src/main/kotlin/com/yonisirote/readmyfeed/tts/TtsVoiceSelection.kt
package com.yonisirote.readmyfeed.tts

private val placeholderVoicePattern = Regex("[-_]language$")

private fun normalizeLanguageTag(value: String): String {
  return value.replace('_', '-').lowercase()
}

private fun getLanguageCandidates(value: String): List<String> {
  val normalized = normalizeLanguageTag(value)
  val segments = normalized.split('-', limit = 2)
  val primary = segments.firstOrNull().orEmpty()
  val region = segments.getOrNull(1)

  // Android devices still expose Hebrew as either he or the older iw tag.
  if (primary == "he" || primary == "iw") {
    return if (region.isNullOrBlank()) {
      listOf("he", "iw")
    } else {
      listOf("he-$region", "iw-$region", "he", "iw")
    }
  }

  return if (region.isNullOrBlank()) {
    listOf(primary)
  } else {
    listOf(normalized, primary)
  }
}

private fun getVoicePriority(voice: TtsVoice): Int {
  val identifier = voice.identifier.lowercase()
  var score = 0

  // Prefer offline, richer voices and heavily penalize placeholder language stubs.
  if (identifier.contains("local")) {
    score += 4
  }

  if (voice.quality == TtsVoiceQuality.ENHANCED) {
    score += 2
  }

  if (identifier.contains("default")) {
    score += 1
  }

  if (voice.requiresNetwork || identifier.contains("network")) {
    score -= 1
  }

  if (placeholderVoicePattern.containsMatchIn(identifier)) {
    score -= 5
  }

  return score
}

private fun pickPreferredVoice(voices: List<TtsVoice>): TtsVoice? {
  var preferredVoice: TtsVoice? = null
  var preferredScore = Int.MIN_VALUE

  for (voice in voices) {
    val score = getVoicePriority(voice)
    if (preferredVoice == null || score > preferredScore) {
      preferredVoice = voice
      preferredScore = score
    }
  }

  return preferredVoice
}

fun findBestVoiceForLanguage(voices: List<TtsVoice>, language: String): TtsVoice? {
  // Try exact locale matches first, then relax to broader language aliases and fallbacks.
  for (candidate in getLanguageCandidates(language)) {
    val matchedVoice = pickPreferredVoice(
      voices.filter { normalizeLanguageTag(it.language) == candidate },
    )

    if (matchedVoice != null) {
      return matchedVoice
    }
  }

  return null
}
